WOC: Feature/vlad/funder dashboard

NedelescuVlad opened the pull request:
<!--
  Thank you for your pull request. Please provide a description above and review
  the requirements below.

  Contributors guide: https://github.com/gitcoinco/web/blob/contributing/CONTRIBUTING.md
-->

##### Description
<!-- A description on what this PR aims to solve -->

WIP - Funder dashboard: https://github.com/gitcoinco/web/issues/1177. Currently have done the front-end with all dummy data. Current state of things is this: https://imgur.com/a/YWirKIn

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [ ] linter status: 100% pass
- [ ] changes don't break existing behavior
- [ ] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, ui, crypto, etc). -->
Frontend
Backend

##### Testing
<!-- Why should the PR reviewer trust that this change doesn't break anything? How have you tested this change? -->

##### Refers/Fixes
<!--
  Link to an issue if applicable. For example:
  If your PR fixes an issue  -> Fixes: #102
  If your PR refers an issue -> Refs: #101
-->

owocki commented:
checking this out now

owocki commented on app/assets/v2/css/funder_dashboard.css:
```diff
@@ -0,0 +1 @@
+/* TODO: Move the inline styles from the html file here
```
yes pls

owocki commented:
this is looking very good and is very exciting to me!

1. just pushed up some merge conflict fixes
2. some responsive design issues http://bits.owocki.com/1o3j2T1j3C3S/Screen%20Shot%202018-06-04%20at%205.37.06%20PM.png http://bits.owocki.com/3k2Q3z3F2V3s/Screen%20Shot%202018-06-04%20at%205.37.35%20PM.png
3. rocketship img is missing http://bits.owocki.com/0s3f3R0g3P3A/Screen%20Shot%202018-06-04%20at%205.38.09%20PM.png
4. 'View Etherscan' should be a link
5. all of the data in the templates is hardcoded into the templates right now.  could you make it so that it's passed down from the views.py file pls?  this will allow the core team to hook up actual data

cc @mbeacom @thelostone-mc @SaptakS for their reviews too!

mbeacom commented on app/dashboard/models.py:
```diff
@@ -85,7 +85,6 @@ def filter_by_status(self, filtered_status=None):
 
 class Bounty(SuperModel):
     """Define the structure of a Bounty.
-
```
This new line is intentional and conforms to pep-257.

Can you re-add this new line?

mbeacom commented on app/dashboard/templates/funder_dashboard.html:
```diff
@@ -0,0 +1,808 @@
+                        {% trans "Reclaim Your Funds" %}
+                      </div>
+                      <div>
+                        You have
```
Can we wrap this and all of the other remaining relevant strings  in `trans` throughout this file ?

mbeacom commented on app/dashboard/views.py:
```diff
@@ -831,6 +831,8 @@ def save_search(request):
     }
     return TemplateResponse(request, 'save_search.html', context)
 
+def funder_dashboard(request):
```
^ Another blank line between `return TemplateResponse(request, 'save_search.html', context)` and `def funder_dashboard(request):`

NedelescuVlad commented:
Thanks for the feedback @owocki, I'm very happy to be working on this as well. I'll be addressing the issues you pointed out. Definitely plenty left to fix and responsive design should have all the love it deserves. In terms of data being spitted out, I assume the goal is to have all the data passed down on GET and make updates with POSTs accordingly (when filtering)?

Just considering this, not sure if what I'm about to suggest is what we want, but - would it be acceptable to render the page with placeholders for data (basically empty divs) and use JS to render the data inside the elements after POST requests? Bear in mind, I am a bit biased towards doing things in JS even when it's not the best solution.

thelostone-mc commented on app/dashboard/templates/funder_dashboard.html:
```diff
@@ -0,0 +1,808 @@
+    }
+
+    .funder-dashboard__header {
+      background-color: #3e00ff;}
```
could we move the bracket  to the next line 
( You've already left a comment on shifting it to an external css file, so I'm good with that )

thelostone-mc commented:
@NedelescuVlad overall looking good :D

- could we change to the font to use `Muli` and the fallback as `sans-serif` ? 

- remove the letter spacing for the numbers
<img width="143" alt="screen shot 2018-06-07 at 3 01 40 pm" src="https://user-images.githubusercontent.com/5358146/41091317-e1900472-6a63-11e8-8730-7ba43ebfe553.png">

- we could reuse the tags from the explorer for this 
<img width="177" alt="screen shot 2018-06-07 at 3 01 29 pm" src="https://user-images.githubusercontent.com/5358146/41091333-e9f99510-6a63-11e8-93b8-93a7f1048509.png">

It's there in `dashboard.html`
<img width="1090" alt="screen shot 2018-06-07 at 3 03 45 pm" src="https://user-images.githubusercontent.com/5358146/41091387-0af66c5c-6a64-11e8-8f68-63ecd90584b6.png">

- Could you update the dropdown to use the existing styles 
checkout `gitcoin/new` you can reuse the code there

<img width="1085" alt="screen shot 2018-06-07 at 3 02 30 pm" src="https://user-images.githubusercontent.com/5358146/41091484-43bd91aa-6a64-11e8-8169-cda7917993f7.png">


- Also this looks weird, the what's next to the avatar.  (just an observation)  :P
<img width="988" alt="screen shot 2018-06-07 at 3 01 59 pm" src="https://user-images.githubusercontent.com/5358146/41091437-279107f0-6a64-11e8-98de-9daba15b78e3.png">


I haven't checked out mobile / tablet view cause you haven't gotten to it. Let me know when I should ^_^

owocki commented:
> Just considering this, not sure if what I'm about to suggest is what we want, but - would it be acceptable to render the page with placeholders for data (basically empty divs) and use JS to render the data inside the elements after POST requests? Bear in mind, I am a bit biased towards doing things in JS even when it's not the best solution.

i'd love to have the data passed down from python into the django templates ( the way gitcoin.co/results works is a great example of this ).

the way you suggested seems like a good way to mock this up / get started.

can't wait for this no longer to be 'WIP'. let me know!

NedelescuVlad commented:
@owocki @thelostone-mc little update on this - I've extracted variables for everything on GET and hooked up the endpoints / documentation for the POST requests, which are for getting the outgoing funds and active bounties. I'm going to fix up some responsive design / UI issues (like the missing rocket ship) next and come back to everything else that still needs to be done, in this order:
- hooking up the links to the issue explorer i.e. view all bounties. Question: where can I find the query params that I can pass to the issue explorer to do the filtering? I've looked a bit but couldn't find them. Will look again. 
- exporting tax report to CSV. How should this work exactly? Imo call an endpoint that returns a csv file? Haven't hooked this up yet
- actual endpoints 

If you guys want to get started on the endpoints at this stage feel free. I will start working on them as well, but will prioritize UI changes first. Hoping to be done with UI in 1 or 2 days max (I can only work on this after work). Thanks guys.

owocki commented:
> hooking up the links to the issue explorer i.e. view all bounties. Question: where can I find the query params that I can pass to the issue explorer to do the filtering? I've looked a bit but couldn't find them. Will look again.

@thelostone-mc could you help here?  to my knoweldge there is not a good way to enter query params for any of the seraches at the moment, but maybe we could built it in.

> exporting tax report to CSV. How should this work exactly? Imo call an endpoint that returns a csv file? Haven't hooked this up yet

for now, just hook up the form to do a POST to a view that stubs out a `message.success()`.  throw a `TODO` in there and i will fill out the code to do this!

> P.S: If you get a js file not found for funder_dashboard.js error just copy the contents of that file in an inline script. I'm getting this error locally and I assume it's because I haven't rebuilt the static assets yet (docker-compose up --build -d doesn't seem to do it?)

i think if you place this file in your `assets/` dir the docker/django setup should automatically serve it from the `static/` dir. ccc  @mbeacom 

> If you guys want to get started on the endpoints at this stage feel free. I will start working on them as well, but will prioritize UI changes first. Hoping to be done with UI in 1 or 2 days max (I can only work on this after work). Thanks guys.

i was thinking of just passing the information into the template as javascript variables .  here's how we do it on the `/results` page => https://github.com/gitcoinco/web/blob/master/app/retail/templates/results.html#L88-L89  either way could work, but this is simpler than providing a whole nother route / endpoint.

owocki commented:
im going away to the mountains for a few days... looking fwd to testing when i get back

thelostone-mc commented:
@owocki @NedelescuVlad 

> hooking up the links to the issue explorer i.e. view all bounties. Question: where can I find the query params that I can pass to the issue explorer to do the filtering? I've looked a bit but couldn't find them. Will look again.

Right now we use localStorage to do the filtering as against query params cause it was getting too long and messy! If we needed we can add that in.

I'm assuming View all bounties -> Show all bounties by the funder.
If that's the case set `bounty_owner_github_username` in localStorage (example: [here](https://github.com/gitcoinco/web/blob/master/app/assets/v2/js/pages/dashboard.js#L225) ) to the funders username

If you wanna add additional filters like Issue Status use `idx_status` ( the list is defined [here](https://github.com/gitcoinco/web/blob/master/app/assets/v2/js/pages/dashboard.js#L11-L21))

PS: remember to invoke `resetFilters` ([here](https://github.com/gitcoinco/web/blob/master/app/assets/v2/js/pages/dashboard.js#L536)) so the previous filters defined by the user don't get applied too


> P.S: If you get a js file not found for funder_dashboard.js error just copy the contents of that file in an inline script. I'm getting this error locally and I assume it's because I haven't rebuilt the static assets yet (docker-compose up --build -d doesn't seem to do it?)

Lemme know once the UI changes are done and I'll check em out and suggest / push changes if needed  ^_^
Placing it under `static` usually ensures that it's picked up

NedelescuVlad commented:
Wasn't able to work on this the past week, was quite busy with family visiting. Thanks for the feedback, that helps a lot

NedelescuVlad commented:
@thelostone-mc  not expecting much more to change in terms of UI. Could you please have a look? My todos at this point are these:

- csv export endpoint
- delete the 2 endpoints so that we pass that data on GET
-  update dropdowns to use the existing styles under gitcoin/new
- make the funder dashboard only clickable when the user is a funder
- total budget input field
- latest bounties -- if no funds to reclaim, box is hidden

OPTIONAL:
- ask about minifying the js / css

thelostone-mc commented:
@NedelescuVlad Awesome !! ^_^
Don't worry about the minifying bit for now -> we've got an PR out at https://github.com/gitcoinco/web/pull/1476 to do the minification 

And the todos look good 👍 
I'll pull in the changes and check it out in a bit

owocki commented:
@mbeacom @thelostone-mc @SaptakS which one of us wants to get cracking on the TODOs that @NedelescuVlad  listed?  im happy to set aside some time when i get back on the grid at home

owocki commented:
UI is looking really good!  cant wait to get real data hooked in!

![screencapture-localhost-8000-funder_dashboard-2018-07-03-13_41_14](https://user-images.githubusercontent.com/513929/42243610-c961021c-7ec6-11e8-80c9-2851dd46c78a.png)

NedelescuVlad commented:
@owocki, I'm gonna look at solving all the todos that I listed this weekend. Looking to do these two first: "delete the 2 endpoints so that we pass that data on GET" and adding a blank csv export endpoint and calling it from the JS / make it download the file. Don't really need help with the TODOs I listed above, but I will need some help in populating all the variables that are extracted in the endpoints. 

Basically I need help coding the real data in, as I haven't spent much time understanding it and the underlying models that you have here at gitcoin.

NedelescuVlad commented:
Question: how can I not display the funder dashboard item in the menu? Is there something to flag if a user is a funder in `document.contxt`? Is there an action in django that is called on every page where I can plug a bit of code in to flag if the user is a funder or not? Or is it fine to let everyone see this link, but for non-funders display a "Sorry you're not a funder" page?

owocki commented:
> Question: how can I not display the funder dashboard item in the menu? Is there something to flag if a user is a funder in document.contxt? Is there an action in django that is called on every page where I can plug a bit of code in to flag if the user is a funder or not? Or is it fine to let everyone see this link, but for non-funders display a "Sorry you're not a funder" page?


please just stub a method called `profile.is_funder` off the profile object and hardcode True for now.  if you throw a TODO in there too @mbeacom and i will update it when we do the data bindings.

NedelescuVlad commented:
Also, just to make sure: In the requirements, under the "All bounties" section
- "Filter ( use/show items from sort menu from above explorer)" does it make sense to use the same filters for the first dropdown as the outgoing funds one? Options are "All, Tip, Payment, Pending, Claimed". Would it not be more useful to be able to filter by bounty status instead? "Open, Stopped, Submitted etc."

mbeacom commented on app/dashboard/models.py:
```diff
@@ -1121,6 +1121,17 @@ def is_staff(self):
         return self.user.is_staff if self.user else False
 
+    @property
+    def is_funder(self):
```
You can determine whether or not the user is classified as a `funder` by checking if `profile.stats[0][0] == 'funder'`

NedelescuVlad replied:
> Ideally, we should split off this role determination to it's own property/method.
Do you mean this should not be a property on profile as we're currently doing?

owocki replied:
> How do I get an instance of the profile in my get method?

`self` ?

> Do you mean this should not be a property on profile as we're currently doing?

i believe he means you shouldnt have to incur the performance over head of `.stats` in order to learn if the user is a funder or hunter

PixelantDesign commented:
Very excited about this!

NedelescuVlad commented:
I've gotten rid of the 2 endpoints and looking to pass that needed data on GET. Have added TODOs in the code where I have questions about the data.

As I'm writing this I noticed that the `Profile` class already has a `get_funded_bounties` method :). How can I get hold of an instance of this class in my code? 

Things left to do:
- Rocket ship icon
- fill in variables with proper data in the back end 
- csv export still
- dropdown style update 
- couple of minor display fixes in the outgoing funds and all bounties table (i.e saying 'Work Started' instead of 'started') 

Questions besides the TODOs:
- Where in the data should I look to see if a bounty is of type 'Tip' or 'Payment'? There is a standard bounty type field in the api response, I thought this may be related but couldn't find any docs. 
- What is a 'Pending' outgoing fund? Has fulfilment even started? (must `fulfillment_started_at != null`m status done and `fulfilment_submitted_at == null`)? How about a `Claimed` one? 
- Are we going to do all this logic in the get method? Feels like these could belong to some domain model class?

@mbeacom @owocki 

Thanks for your time and I know this is something you're all looking forward to getting out. I'll try putting in some work during the week as well and get this tidied up by the end of the week. But yeah, my main blocker right now is coding the back-end. 

P.S: @PixelantDesign do you happen to still have that cool rocket icon from the design? No worries if not, I can make it again in GIMP (think I found something very similar on google a while back)

NedelescuVlad commented:
Starting to understand how to use python classes :)

owocki commented:
> As I'm writing this I noticed that the Profile class already has a get_funded_bounties method :). How can I get hold of an instance of this class in my code?

request.user.profile

> Where in the data should I look to see if a bounty is of type 'Tip' or 'Payment'? There is a standard bounty type field in the api response, I thought this may be related but couldn't find any docs.

Bounties are in the `Bounty` object.
Tips are in the `Tip` object.

> What is a 'Pending' outgoing fund? Has fulfilment even started? (must fulfillment_started_at != nullm status done and fulfilment_submitted_at == null)? How about a Claimed one?

I'm not sure what model you're talking about here.

A claimed one will have `bounty.interests.exists()`

> Double checking: Do we want to filter all bounties by the same things we filter outgoing funds by? Does it make sense to filter by type 'Tip'/'Payment' where the type displayed is something like 'Front-end'? I might have misunderstood this so please confirm.

I don't have context on what you're talking about.  Might be good to ask inline on the code.

> Are we going to do all this logic in the get method? Feels like these could belong to some domain model class?

Again, I think it's better to move this convo to the code.  I dont know which get method you're talking about.

NedelescuVlad commented:
:+1: will leave comments in the code. 
> Are we going to do all this logic in the get method? Feels like these could belong to some domain model class?
Discard that comment, am a bit more informed now

owocki commented:
very excited to see the data starting to be added here!

vlad, let us know if you need input here.  i'm happy to extend some more funds here too since the scope has mushroomed with the addition of backend stuff

NedelescuVlad commented:
@owocki Thank you, but I'll kindly say no to the extra funds. If anyone wants to offer some inline comments in terms of the TODOs I'd be happy to have those though.

owocki commented:
commenting inline on TODOs now

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    funder_bounties = request.user.profile.get_funded_bounties()
+
+    # TODO: Remove, it's just for testing now
+    funder_bounties = Bounty.objects.all()
```
youll want to filter on `current_bounty=True` here

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    done_bounties = active_done_expired_bounties.filter_by_status(['done'])
+    expired_bounties = active_done_expired_bounties.filter_by_status(['expired'])
+
+    contributor1 = {
```
this looks hardcoded

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    }
+
+    # TODO: use done bounties to get the contributors
+    top_contributors = [contributor1]
```
this is easy to do with the `bounty.bounty_owner_github_username` param

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    # Each one of these dictionaries is composed of 2 arrays that should have the same length.
+    # Each data[i] labels[i] represents a X Y coordinate (plot point)
+    payout_history_weekly = {
+        "data": [50, 105, 405, 150, 200, 300, 80, 305],
```
you can use the `bounty.fulfillment_accepted_on` field to figure out the payout over time

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    }
+
+    payout_history_monthly = {
+        "data": [80, 305, 50, 105, 405, 150, 200, 300],
```
same comment as above

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    }
+
+    payout_history_yearly = {
+        "data": [50000, 70000, 90000, 30000],
```
+1

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+
+    expiring_days_count = last_expiring_days_from_now
+
+    # TODO: Count the new comments
```
`bounty.github_comments`

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    submitted_bounties_count = funder_bounties.count()
+
+    # TODO: number of people who ever contributed
+    total_contributors_count = 142
```
you can do this by summing up all the `current_bounty=True`s that belong to this user and counting all of the fulfillments that are accepted

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+            total_paid_eth = total_paid_eth + bounty.get_value_in_eth
+
+    # TODO: total paid date since - date of first submitted bounty
+    total_paid_date_since = _("May 5. 2018")
```
this is fairly simple to do by filtering on the bounties and gettin their `web3_created_on` date

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    total_budget_eth = ""
+
+    # TODO: Tax year stats - Considering the tax year to be 1 Jan to 31 Dec
+    tax_year = 2018
```
you might want to default this to this year if > march 31st, last year if not

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    # TODO: Tax year stats - Considering the tax year to be 1 Jan to 31 Dec
+    tax_year = 2018
+    tax_year_bounties_count = 139
+    tax_year_bounties_worth_dollars = 34.500
```
you can use `bounty.fulfillment_accepted_on` to query this data

owocki replied:
`bounty.value_in_usdt`

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    tax_year_bounties_worth_dollars = 34.500
+
+    # Latest on your bounties
+    # TODO: What's the difference between an 'issue' and a 'bounty'? Just using bounty for these right now..
```
they are the same

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+
+    outgoing_funds = []
+    for bounty in done_bounties.filter(fulfillment_started_on__isnull=False, fulfillment_submitted_on__isnull=True):
+        # TODO: Need the txid to generate this link. Where is it in the bounty object, if at all?
```
https://github.com/gitcoinco/web/pull/1722 

adds the ability to pull txids.  you can expect that on a go-forward basis the txids will be there

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+        # TODO: Need the txid to generate this link. Where is it in the bounty object, if at all?
+        etherscan_link = '#'
+
+        # TODO: Should we check if the user has received the bounty instead, similar to a Tip's received_on field?
```
yes.  if the user is in `bounty.fulfillments.filter(accepted=True)` then they have received the bounty

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+        if tip.status == "PENDING":
+            etherscan_link = "https://etherscan.io/tx/" + tip.txid
+            outgoing_funds.append({
+                'id': 'TODO: Tip id',  # TODO: Get issue number that the tip is for
```
`tip.bounty`

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+            etherscan_link = "https://etherscan.io/tx/" + tip.txid
+            outgoing_funds.append({
+                'id': 'TODO: Tip id',  # TODO: Get issue number that the tip is for
+                'title': bounty.title, # TODO: Get issue name that the tip is for
```
tip.bounty.title

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+
+        # TODO: Should we check if the user has received the bounty instead, similar to a Tip's received_on field?
+        # if bounty.interests.exists(): # TODO: interests is not defined, am I using it right?
+        if False:
```
`bounty.interested.all()` not `interests`

owocki commented on app/dashboard/templates/shared/nav_auth.html:
```diff
@@ -43,6 +43,13 @@
         <i class="fas fa-user"></i>
         {% trans "View Profile" %}
       </a>
+      <!-- TODO: Funder dashboard check if user is funder -->
```
is this still TODO?

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+        # if bounty.interests.exists(): #  TODO: interests is not defined, am I using it right?
+            pending_or_claimed = 'Claimed'
+        else:
+            # TODO: Is any bounty that's not claimed, in pending status?
```
if status is in `open`, `started`, or `submitted` it is pending

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+            pending_or_claimed = 'Pending'
+
+        # TODO: Is the bounty a Tip or a Payment?
+        tip_or_payment = None
```
whats the TODO asking here?

owocki commented on app/dashboard/views.py:
```diff
@@ -924,6 +927,282 @@ def save_search(request):
+    all_bounties = []
+    for bounty in funder_bounties:
+        if False:
+        # if bounty.interests.exists(): #  TODO: interests is not defined, am I using it right?
```
`bounty.interested.all()`

NedelescuVlad commented:
Gave a confused emoji on your comment for "this is easy to do with the `bounty.bounty_owner_github_username` param" - I think I should use `fulfiller_github_username` rather?

NedelescuVlad commented:
Going to add a few more inline comments for where I still have questions. Would be great if one of you guys can test this out when you get the time with your own accounts or something and make suggestions. I haven't tested the outgoing funds table as I can't get anything to show for now - will look more into it although it might just be a test data thing. 

I still have the csv export to do and solve merge issues / linter / address your comments. This looks like it can go out soon if my backend hasn't been too shabby :)

owocki commented:
i get an exception when i try to load the page

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/wsgi.py", line 146, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 81, in get_response
    response = self._middleware_chain(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/whitenoise/middleware.py", line 25, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/utils/deprecation.py", line 95, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 37, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 87, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 122, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.6/site-packages/django_extensions/management/technical_response.py", line 6, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/code/app/dashboard/views.py", line 959, in funder_dashboard
    payout_history = get_payout_history(done_bounties)
  File "/code/app/dashboard/helpers.py", line 732, in get_payout_history
    m[month] = m[month] + bounty_val
TypeError: unsupported operand type(s) for +: 'float' and 'NoneType'
```

can we resolve the merge conflicts / merge the latest master in here?  this is based off a pretty old branch!

NedelescuVlad commented:
@owocki that issue should be solved now.

NedelescuVlad commented:
Have just added csv export and monthly / quarterly budget (though that's not posting to the server and is outputting to console).

One thing I'd like to add is pagination for the bounties / outgoing funds.  (right now I'm slicing 5 items out of each in JS, planning to commit that change when I come back in a few hours). Also gonna look at fixing the linter errors :) 

Also, if you guys can also take a look at the `activateLinksToIssueExplorer` method inside `funder_dasbhoard. js` it would be great :) 

Looking forward to your feedback/suggestions.

NedelescuVlad commented:
Hopefully everything builds fine now 😬

mbeacom commented:
Starting to review this...  A few quick notes:

Failing eslint:  

```app/assets/v2/js/pages/funder_dashboard.js
  344:14  error  'getBounties' is already defined  no-redeclare
```

Failing stylelint:

```
app/assets/v2/css/funder_dashboard.css
 225:3  ✖  Expected selector "tr" to come before selector "thead tr"   no-descending-specificity
```

NedelescuVlad commented:
Have added pagination for all bounties, might still be a bit buggy

owocki commented:
testing now

owocki commented:
just pushed up some changes to get this to run on my local 13d9cbddad7e1bf60f01878aa59caea41251475b and left somme comments inline.

Vlad, I'm very impressed at your tenacity on this feature. can't wait to get it across the finish line and payout!  i'll throw in some more ETH since it's taken a while and you did the backend too.

1. [these values](http://bits.owocki.com/2R0a353h2M19/Screen%20Shot%202018-08-08%20at%206.23.46%20PM.png) are expressed in wei, but really they should be expressed in whole units of ETH.  
1a. same with the ETH value [here](http://bits.owocki.com/28001H1t1944/Screen%20Shot%202018-08-08%20at%206.26.47%20PM.png)
1b. looks like the 'download report' button has this problem too!
2. can we change the page title to 'Funder Dashboard'
3. the 'top contributors' profile links are broken.  they should go to `gitcoin.co/profile/[handle]`
4. my 'active bounties' count is wrong http://bits.owocki.com/151g433U0e1O/Screen%20Shot%202018-08-08%20at%206.26.01%20PM.png
5. for the ['budget' module](http://bits.owocki.com/3m1j3X0Z3w2T/Screen%20Shot%202018-08-08%20at%206.26.19%20PM.png), maybe we should just let you set a budget and save it in the backend? 
6. [this number](http://bits.owocki.com/0u3v1b3i0X2S/Screen%20Shot%202018-08-08%20at%206.27.18%20PM.png) is wrong.  please make sure you're filtering only on `Bounties.object.current()`
7. i think [this module](http://bits.owocki.com/3h0X2C371U0d/Screen%20Shot%202018-08-08%20at%206.27.46%20PM.png) needs a little TLC.  perhaps we should hide it if we don't have anything good to say.   otherwise, maybe it can just say you rank on the funder leaderboard?

i dont want the length of the list above to detract from the fact that this *has come a really long way and im thrilled about it!!*

@mbeacom any thoughts here?

NedelescuVlad commented:
Thanks for the comments kevin, looking at solving these. Will keep you posted

NedelescuVlad commented:
Also, was the outgoing funds module working fine for you? That would be the table. Having a hard time testing it since I don't have any outgoing funds :)

owocki commented:
am not by a computer to test... but i'd suggest opening up a bounty or sending a tip on rinkeby (checkout https://faucet.rinkeby.io/ for RETH) if you want to test outgoing funds!

thelostone-mc commented:
@NedelescuVlad could you give us a date by when this would be ready ? If you need test ETH & unable to use the rinkeby faucet -> I'd be happy to send some across

Would like to ship this out as soon as possible ^_^

NedelescuVlad commented:
@thelostone-mc I'm back this weekend. Do check up on me in 24 hours. Going to work on this tomorrow and expect it to be done by 10 PM UTC time.

NedelescuVlad commented:
The end is in sight. Going to make the total budget module post to the server tomorrow and do a lot more testing. Also going to make the missing etherscan link on Bounty be handled gracefully. 

Was able to use the rinkeby faucet, although I haven't tested since I've switched to using a new PC (currently running `docker-compose exec web python3 app/manage.py sync_geth rinkeby 402 99999999999` which I'll be leaving overnight).

NedelescuVlad commented:
@owocki @thelostone-mc can we confirm what filter values we want for the bounties module? https://i.imgur.com/Kmr4Qdw.png

NedelescuVlad commented:
One more thing, when comparing USD to ETH it's not at the current conversion rate. I'm currently using `get_value_in_eth`, should I switch to `get_value_in_eth_then` or `get_value_in_eth_now` if any?

NedelescuVlad commented:
Ah, still have to do the total budget module posting to the server. Unsure where to save this data though

owocki commented:
> One more thing, when comparing USD to ETH it's not at the current conversion rate. I'm currently using get_value_in_eth, should I switch to get_value_in_eth_then or get_value_in_eth_now if any?

`get_value_in_eth` should be fine

owocki commented:
just QAed this... its really coming together!   i'm really blown away by all the persistent progress

further notes:

1. for the 'view'  and 'view all payments' buttons, itd be great to have those enabled as 'deep links', per https://github.com/gitcoinco/web/pull/2062
2. for 'outgoing funds', if the etherscan link is unavailable, could we link the user to the invoice for those outgoing funds?  `bounty.action_urls['invoice']` should do it
3. the '0 ACTIVE BOUNTIES' portion seems to be working.. i pushed a fix to this to the branch
4. for 'total budget' can we enable this such that it'll just save the budget to the database, and re-display it, for now?  this will allow us to know how much budget users have.
5. i dont think ' contributors have commented on your open issue' makes sense to show (as those are not all new comments', so i am removing for now. we can reintroduce this porttion when https://github.com/gitcoinco/web/issues/2050 goes live
6. i dont think the csv file is properly escaped [example](http://bits.owocki.com/08320fd77e87/GitcoinPaidBountiesThisYear%20(1).csv) as columns are bleeding into each other.

PixelantDesign commented:
@NedelescuVlad We're very excited about this feature! Would it be possible to wrap this up this weekend? I'll Tip Extra for this! Thanks!

NedelescuVlad commented:
Just saw your comment and kevin's! Gonna look to wrap this up this weekend

thelostone-mc commented:
@NedelescuVlad Just an update!
I'll be working on the contributor dashboard and since you've got pieces of code I can reuse 
I'm gonna shamelessly refactor it in and reuse code  :D

NedelescuVlad commented:
Have points 1 and 4 from kevin's that i have left to do. It's a bank holiday in the UK so i have tomorrow off as well and will be solving these.

NedelescuVlad commented:
@owocki @PixelantDesign @thelostone-mc can I get one more QA test? Have just pushed the last item on that todo list (adding total budget submission)

NedelescuVlad commented:
This 'method is too complex' warning, shall I refactor some of the code into helpers/utils?

owocki commented:
yes please, if it's not too much work!

On Mon, Aug 27, 2018 at 5:50 PM, Vlad <notifications@github.com> wrote:

> This 'method is too complex' warning, shall I refactor some of the code
> into helpers/utils?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitcoinco/web/pull/1351#issuecomment-416380526>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfXicFf94c-yfxuzaU-zEEHUBXW9SScks5uVGmigaJpZM4UYOKi>
> .
>



-- 


@owocki <http://www.twitter.com/owocki>

_________________________________________

check out what i'm building on github <http://github.com/owocki>

or what i'm shooting on photoshelter <http://photography.owocki.com/>

or find me on twitter <http://www.twitter.com/owocki>, facebook
<http://www.facebook.com/kevinowocki>, instagram
<http://instagram.com/owocki>, linkedin <http://www.linkedin.com/in/owocki>,
and owocki.com <http://www.owocki.com/?utm_source=emailsig>.

see you around teh internets!

PS -- Come to the next Boulder Blockchain
<https://www.meetup.com/Boulder-Blockchain/> meetup.  Be there or be ^ 2.
PPS -- Have I mentioned that *G**itcoin is live?*  More @ https://gitcoin.co

NedelescuVlad commented:
@owocki shouldn't be too much work, will try to be careful when refactoring

NedelescuVlad commented:
@owocki @thelostone-mc I've just finished refactoring - currently getting a travis build error due to imports not being in the correct order (isort) and stickler errors from having merged master into my branch 
Lmk when you can give this another qa look

thelostone-mc commented:
<img width="819" alt="screen shot 2018-09-07 at 12 59 46 am" src="https://user-images.githubusercontent.com/5358146/45180566-65775700-b239-11e8-909b-62231b5368d2.png">

The sizing / text alignment seems off ! could we fix that up ?

NedelescuVlad commented:
Sure, do you refer to the grey footers not being the same height? Once
users give a total budget these should line up nicely. Or is it that the
entire height seems stretched / top text is not aligned?

On Thu, 6 Sep 2018, 22:31 Aditya Anand M C, <notifications@github.com>
wrote:

> [image: screen shot 2018-09-07 at 12 59 46 am]
> <https://user-images.githubusercontent.com/5358146/45180566-65775700-b239-11e8-909b-62231b5368d2.png>
>
> The sizing / text alignment seems off ! could we fix that up ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitcoinco/web/pull/1351#issuecomment-419214368>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AOSj8KQklYDWFj8yZmoBuTec_xra16ejks5uYXgCgaJpZM4UYOKi>
> .
>

mbeacom commented:
@NedelescuVlad That pip output are merely warnings.  We have package upgrades (specifically web3/eth related) on our radar. Thanks for relaying!

thelostone-mc commented:
@NedelescuVlad 

> Or is it that theentire height seems stretched / top text is not aligned?

^ That

NedelescuVlad commented:
@thelostone-mc just pushed a quick fix, this is the new layout https://i.imgur.com/BcKgNSX.png ... I don't think I can do much about the stretching when the total budget module is in the input status as it takes quite a lot of space there. 

A workaround for that would be imo to let users input a total budget in a modal and provide just a text link on the page that pops up the modal... though it's also something that can be reworked later on i think

thelostone-mc commented:
> A workaround for that would be imo to let users input a total budget in a modal and provide just a text link on the page that pops up the modal... though it's also something that can be reworked later on i think

Yup! we can look into that later.

LGTM from code and what I locally tested.

@mbeacom / @SaptakS can we put this up on staging?

mbeacom commented on app/dashboard/templates/funder_dashboard.html:
```diff
@@ -0,0 +1,667 @@
+                              data-is-status-pending-or-claimed-filter="true"
+                              {% endif %}
+                              >
+                        {{filter.value_display}}
```
Can we maintain consistency with `{{ var }}` vs `{{var}}`

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,188 @@ def get_quickstart_video(request):
+@cache_page(60 * 60 * 24)
+def funder_dashboard(request):
+    """ Render the funder dashboard"""
+
```
No new line after docstring. No space before starting docstring.

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1144,6 +1153,188 @@ def get_quickstart_video(request):
+def funder_dashboard(request):
+    """ Render the funder dashboard"""
+
+    if not hasattr(request.user, 'profile') or not request.user.profile.is_funder:
```
We shouldn't restrict this view to only `is_funder` types.

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,188 @@ def get_quickstart_video(request):
+
+    funder_bounties = request.user.profile.get_funded_bounties()
+
+    active_done_expired_bounties = funder_bounties.filter_by_status(['active', 'done', 'expired'])
```
Should this be using predefined lists of status types?

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,188 @@ def get_quickstart_video(request):
+        if issue_worth_in_usdt is not None:
+            d_expired_issues_worth_dollars = float(d_expired_issues_worth_dollars) + float(issue_worth_in_usdt)
+
+    d_active_bounties_count = active_bounties.count()
```
If we're using a variable in only one place, can we omit the variable assignment and use whatever the variable is calling in line?  In this case, `active_bounties.count()` could be on line 1292

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,188 @@ def get_quickstart_video(request):
+@csrf_exempt
+@require_POST
+def update_funder_total_budget(request):
+    data = json.loads(request.body)
```
Docstring

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,188 @@ def get_quickstart_video(request):
+
+    profile = request.user.profile
+
+    profile.funder_total_budget_usdt = budget_usd
```
You could update this to use `profile.update(funder_total_budget_usdt=budget_usd, ...)`

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,188 @@ def get_quickstart_video(request):
+
+    profile.save()
+
+    result = {
```
Do we need this assignment? Could we just do: `return JsonResponse({'status': 200})` if no alternative response occurs?

mbeacom commented on app/economy/utils.py:
```diff
@@ -92,3 +92,11 @@ def etherscan_link(txid):
+
+def eth_from_wei(wei_amount):
+    """ Converts a wei amount to a whole value eth amount """
+
```
No new line, no space in the beginning, end in punctuation, imperative mood.

mbeacom commented on app/dashboard/views.py:
```diff
@@ -1141,6 +1151,187 @@ def get_quickstart_video(request):
 
+@vary_on_cookie
+@cache_page(60 * 60 * 24)
+def funder_dashboard(request):
```
Any chance we can DRY up everything newly introduced throughout the PR, make variables a little bit more descriptive, and maybe add some comments and/or adjust some of the loop logic to make maintainability/readability easier?

mbeacom commented:
@NedelescuVlad Thanks for the update.  Can you look into the comments I left? Feel free to let me know if you have any questions.

NedelescuVlad commented:
@mbeacom I've just pushed a commit addressing the bulk of your comments. The remaining ones are:
https://github.com/gitcoinco/web/pull/1351#discussion_r216677497 (yup, working on it)
https://github.com/gitcoinco/web/pull/1351#discussion_r216680067 (can do, but i'm using more fields than what that method returns so it's not a straight forward mapping)
https://github.com/gitcoinco/web/pull/1351#discussion_r216685464 (mostly done, haven't resolved it yet due to the comment above)

mbeacom commented:
@NedelescuVlad Regarding `to_standard_dict()`:

You can assign the response of `to_standard_dict()` to a var and perform your additions like:

```python
def to_funder_dashboard_bounty(bounty):
    bounty_dict = bounty.to_standard_dict(fields=['bounty_type', 'status', 'github_url'])
    bounty_dict.update({..., 'pending_or_claimed': 'None', ...})

    if bounty.interested.exists():
        bounty_dict['pending_or_claimed'] = 'Claimed'
    if bounty.status in ['open', 'started', 'submitted']:
        bounty_dict['pending_or_claimed'] = 'Pending'

return bounty_dict
```
One thing I'm noticing here, though...  When do we want to show claimed? In this case, if there are no submissions and someone has conveyed interest `Start work` (i.e. `bounty.interested.exists() == True`), it's always going to be updated to `Pending` regardless of the existence of interest objects because the status will be `started`.

NedelescuVlad commented:
> @NedelescuVlad Regarding `to_standard_dict()`:
> 
> You can assign the response of `to_standard_dict()` to a var and perform your additions like:
> 
> ```python
> def to_funder_dashboard_bounty(bounty):
>     bounty_dict = bounty.to_standard_dict(fields=['bounty_type', 'status', 'github_url'])
>     bounty_dict.update({..., 'pending_or_claimed': 'None', ...})
> 
>     if bounty.interested.exists():
>         bounty_dict['pending_or_claimed'] = 'Claimed'
>     if bounty.status in ['open', 'started', 'submitted']:
>         bounty_dict['pending_or_claimed'] = 'Pending'
> 
> return bounty_dict
> ```
> 
> One thing I'm noticing here, though... When do we want to show claimed? In this case, if there are no submissions and someone has conveyed interest `Start work` (i.e. `bounty.interested.exists() == True`), it's always going to be updated to `Pending` regardless of the existence of interest objects because the status will be `started`.

Good catch on the Claimed/Pending status, thank you very much! I've used an elif for the 'Pending' branch. The snippet for `to_standard_dict()` makes sense and is a lot cleaner. I will be renaming some of the keys though as they are tightly coupled to the front-end

NedelescuVlad commented:
Need to squash my commits, merge master into this again and fix the migration conflicts. Otherwise looks good imo.

One concern I have is regarding to `@cache_page` + `@vary_on_cookie` combination. Would it be possible for a user that has the same cookies as another to see that other user's funder dashboard? Would it ever be the case that 2 users have the same cookies, if they are logged in? (I noticed there's a sessionid cookie, is that set per user?)

NedelescuVlad commented:
@mbeacom Just finished squashing my commits. I ended up squashing some commits from other contributors to this PR as well... sorry for that but didn't know how to get around it

mbeacom commented:
@NedelescuVlad I hate to ask this... but can we maybe raise this in a separate branch or resolve the commit history overwrite?  I don't want to muck up the commit history on `master`.  Plus if you simply make a patch file from your branch, it'll squash your additional commits.

NedelescuVlad commented:
@mbeacom Awesome, didn't know about patch files, that's exactly what i need. Will bring this to how it was and look into using a patch file instead

NedelescuVlad commented:
> @NedelescuVlad I hate to ask this... but can we maybe raise this in a separate branch or resolve the commit history overwrite? I don't want to muck up the commit history on `master`. Plus if you simply make a patch file from your branch, it'll squash your additional commits.

Could you have a look here please when you get the time? https://github.com/gitcoinco/web/compare/master...NedelescuVlad:feature/vlad/funder-dashboard-v3-throwaway?expand=1 

I've squashed a bunch of commits into that "Add funder dashboard" commit but am unsure what to do with the ones that belong to other contributors. Am I fine to also squash those since they aren't on master at the moment anyway?

I'm thinking that when we're happy with how the history looks on that branch I'll replace this one's history with that one's

PixelantDesign commented:
Hey @NedelescuVlad Thanks for all of the hard work! We'd like to take this back in and push the V1 out. **Please click submit work** so that you can get paid!

NedelescuVlad commented:
@PixelantDesign thank you very much and sorry I couldn't get this further :) Good luck with tidying up

PixelantDesign commented:
Hey @NedelescuVlad It looks like some of the data points are displaying on the dashboard, are you able to help take a quick look?

NedelescuVlad commented:
Yup, could you send a screenshot please? Either here or you can find me on
slack as nedelescuvlad or vladnedelescu

On Tue, 18 Sep 2018, 19:33 Pixelant, <notifications@github.com> wrote:

> Hey @NedelescuVlad <https://github.com/NedelescuVlad> It looks like some
> of the data points are displaying on the dashboard, are you able to help
> take a quick look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitcoinco/web/pull/1351#issuecomment-422461259>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AOSj8HtrNp8vAf7oN8Z5e514FHwMkmSMks5ucSB0gaJpZM4UYOKi>
> .
>

NedelescuVlad commented:
@thelostone-mc could we close this PR and leave a link to the new PR this got moved to?

thelostone-mc commented:
@NedelescuVlad I'm okay with that. Moving this over to https://github.com/gitcoinco/web/pull/2239
